### ISSUE TITLE
SC3026: shellcheck does not report that [^a] is not equivalent to [!a] in dash

### ISSUE THREAD

**Issue by emanuele6** [koalaman/shellcheck#2361]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC3026
- My shellcheck version (`shellcheck --version` or "online"): online
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/dash

: >> a
: >> b
: >> c

for m in [^a]
  do printf 'match: %s\n' "$m"
done
# output:
# match: a
```

#### Here's what shellcheck currently says:

*nothing*

#### Here's what I wanted or expected to see:

```
Line 7:
for m in [^a]
         ^-- SC3026 (warning): In dash, [^a] is not equivalent to [!a].
```

---

If the shebang is `#!/bin/sh`, shellcheck will report the folowing:
```
Line 7:
for m in [^a]
         ^-- SC3026 (warning): In POSIX sh, ^ in place of ! in glob bracket expressions is undefined.
```

I was going to report that `shellcheck` doesn't warn about this in `dash`, but after reading the source code

https://github.com/koalaman/shellcheck/blob/5669eb22037980c5a6b74b0d420cb452990bcf88/src/ShellCheck/Checks/ShellSupport.hs#L238-L239

I figured this was probably intentionally not being reported since the behaviour of `[^a]` in `dash` is defined and is equivalent to `[a^]` in other shells.

But then I read [the wiki page for SC3026](https://github.com/koalaman/shellcheck/wiki/SC3026) and saw:

> `[^c]` is frequently used in most regular expression variants to mean "any character except `c`". This is so pervasive that bash, ksh, dash, and BusyBox ash, all allow it.

All the mentioned shells, except `dash`, allow `[^a]` as an alternative to `[!a]`.

So I realised that, maybe, `dash` was being ignored because who wrote it incorrectly thought that `dash` supported it.

I am reporting the issue since, at least, the wiki page is incorrect.

If we want to keep not reporting `[^a]` for dash, I think that the wiki page should at least mention that if the intent of the glob was to match "any character that is either `a` or `^`", `[a^]` should be preferred to `[^a]`.

---

Here is a script to check the behaviour of `[!a]` vs `[^a]` vs `[a^]` in different POSIX shells:

```sh
#!/bin/sh

IFS=

for f in a b c '^'
  do : >> "$f"
done

for pat in '[!a]' '[^a]' '[a^]'; do
  for i in $pat
    do printf '%s matches: %s\n' "$pat" "$i"
  done
done
```

<details><summary>Output for <code>bash</code>, <code>ksh</code>, <code>dash</code> and <code>busybox sh</code></p></summary><p>

```
$ bash script.sh
[!a] matches: ^
[!a] matches: b
[!a] matches: c
[^a] matches: ^
[^a] matches: b
[^a] matches: c
[a^] matches: ^
[a^] matches: a
$ ksh script.sh
[!a] matches: ^
[!a] matches: b
[!a] matches: c
[^a] matches: ^
[^a] matches: b
[^a] matches: c
[a^] matches: ^
[a^] matches: a
$ dash script.sh
[!a] matches: ^
[!a] matches: b
[!a] matches: c
[^a] matches: ^
[^a] matches: a
[a^] matches: ^
[a^] matches: a
$ busybox sh script.sh
[!a] matches: ^
[!a] matches: b
[!a] matches: c
[^a] matches: ^
[^a] matches: b
[^a] matches: c
[a^] matches: ^
[a^] matches: a
```

<details><summary>Shell versions</summary><p>

```
$ bash -c 'declare -p BASH_VERSINFO'
declare -ar BASH_VERSINFO=([0]="5" [1]="1" [2]="8" [3]="1" [4]="release" [5]="x86_64-pc-linux-gnu")
$ ksh -c 'typeset -p .sh.version'
.sh.version='Version A 2020.0.0'
$ pacman -Qi dash | grep '^Version'
Version         : 0.5.11.4-1
$ pacman -Qi busybox | grep '^Version'
Version         : 1.33.1-1
```
</p></details>
</p></details>
